drivers/sensor/ti/ina23x/ina230.c

```diff
@@ -29,41 +29,65 @@ static int ina230_channel_get(const struct device *dev, enum sensor_channel chan
 {
 	struct ina230_data *data = dev->data;
 	const struct ina230_config *const config = dev->config;
-	uint32_t bus_uv, power_uw;
-	int32_t current_ua;
+	int32_t val_micro;
 
 	switch (chan) {
 	case SENSOR_CHAN_VOLTAGE:
-		bus_uv = data->bus_voltage * config->uv_lsb;
-
-		/* convert to fractional volts (units for voltage channel) */
-		val->val1 = bus_uv / 1000000U;
-		val->val2 = bus_uv % 1000000U;
+		val_micro = data->bus_voltage * config->uv_lsb;
 		break;
-
 	case SENSOR_CHAN_CURRENT:
-		/* see datasheet "Programming" section for reference */
-		current_ua = data->current * config->current_lsb;
-
-		/* convert to fractional amperes */
-		val->val1 = current_ua / 1000000L;
-		val->val2 = current_ua % 1000000L;
+		val_micro = data->current * config->current_lsb;
 		break;
-
 	case SENSOR_CHAN_POWER:
-		power_uw = data->power * config->power_scale * config->current_lsb;
-
-		/* convert to fractional watts */
-		val->val1 = (int32_t)(power_uw / 1000000U);
-		val->val2 = (int32_t)(power_uw % 1000000U);
-
+		val_micro = data->power * config->power_scale * config->current_lsb;
 		break;
-
 	default:
 		return -ENOTSUP;
 	}
 
-	return 0;
+	return sensor_value_from_micro(val, val_micro);
+}
+
+static int ina230_adc_run(const struct device *dev)
+{
+	const struct ina230_config *config = dev->config;
+	bool in_shutdown = config->adc_mode == 0;
+	uint16_t reg;
+	int ret;
+
+	/* Modes 5 through 7 are continuously sampling */
+	if (config->adc_mode >= 5) {
+		return 0;
+	}
+
+	/* If idling in shutdown, update to single shot triggered */
+	reg = in_shutdown ? (config->config | 3) : config->config;
+
+	/* Write the configuration register to force a sample */
+	ret = ina23x_reg_write(&config->bus, INA230_REG_CONFIG, reg);
+	if (ret < 0) {
+		LOG_ERR("Failed to trigger sample");
+		return ret;
+	}
+
+	/* Wait until we expect the sampling to be complete */
+	k_sleep(K_USEC(config->conv_duration_us));
+
+	/* Poll conversion complete bit */
+	while (1) {
+		ret = ina23x_reg_read_16(&config->bus, INA230_REG_MASK, &reg);
+		if ((ret != 0) || (reg & INA230_REG_MASK_CNVR)) {
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	}
```

Comment:
This can loop infinitely, use:

```c
WAIT_FOR(
    ina23x_reg_read_16(&config->bus, INA230_REG_MASK, &reg) != 0 || (reg & INA230_REG_MASK_CNVR),
    K_MSEC(5), /* Use whatever reasonable timeout you think is right here */
    k_msleep(1));
```

Reply:
Cool macro but it doesn't work for this use case.
Read failures will just break the loop with no error code, and you don't know whether the bit is set or not without testing it again outside the loop.